examples/vision/patch_conv_net.py

```diff
@@ -0,0 +1,661 @@
+convnets. They propose to substitute the global average pooling layer
+of a convnet with a transformer layer. The self-attention layer of the
+transformer would produces the attention maps that corresponds to the
+most attended patches of the image for the classification decision.
+
+In this example, we are going to minimally implement the ideas of
+[Augmenting Convolutional networks with attention-based aggregation](https://arxiv.org/abs/2112.13692).
+The main motive of this example is to cover the following ideas with
+minor modifications (to adjust the implementation with CIFAR10):
+
+- The simple design for the attention-based pooling layer, such that
+    it explicitly provides the weights (importance) of the different
+    patches.
+- The novel architecture of convnet called the **PatchConvNet** which
+    deviates from the age old pyramidal architecture.
+"""
+
+"""
+## Setup and Imports
+
+This example requires TensorFlow Addons, which can be installed using
+the following command:
+
+```shell
+pip install -U tensorflow-addons
+```
+"""
+
+import math
+import requests
+import numpy as np
+from PIL import Image
+import tensorflow as tf
+from tensorflow import keras
+import matplotlib.pyplot as plt
+import tensorflow_addons as tfa
+from tensorflow.keras import layers
+
+# Set seed for reproducibiltiy
+SEED = 42
+keras.utils.set_random_seed(SEED)
+
+"""
+## Hyperparameters
+"""
+
+# DATA
+BATCH_SIZE = 128
+BUFFER_SIZE = BATCH_SIZE * 2
+AUTO = tf.data.AUTOTUNE
+INPUT_SHAPE = (32, 32, 3)
+NUM_CLASSES = 10  # for CIFAR 10
+
+# AUGMENTATION
+IMAGE_SIZE = 48  # We will resize input images to this size.
+
+# ARCHITECTURE
+DIMENSIONS = 256
+SE_RATIO = 8
+TRUNK_DEPTH = 2
+
+# OPTIMIZER
+LEARNING_RATE = 1e-3
+WEIGHT_DECAY = 1e-4
+
+# PRETRAINING
+EPOCHS = 50
+
+"""
+## Load the CIFAR10 dataset
+"""
+
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+(x_train, y_train), (x_val, y_val) = (
+    (x_train[:40000], y_train[:40000]),
+    (x_train[40000:], y_train[40000:]),
+)
+print(f"Training samples: {len(x_train)}")
+print(f"Validation samples: {len(x_val)}")
+print(f"Testing samples: {len(x_test)}")
+
+train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+train_ds = train_ds.shuffle(BUFFER_SIZE).batch(BATCH_SIZE).prefetch(AUTO)
+
+val_ds = tf.data.Dataset.from_tensor_slices((x_val, y_val))
+val_ds = val_ds.batch(BATCH_SIZE).prefetch(AUTO)
+
+test_ds = tf.data.Dataset.from_tensor_slices((x_test, y_test))
+test_ds = test_ds.batch(BATCH_SIZE).prefetch(AUTO)
+
+"""
+## Augmentation Layers
+"""
+
+
+def get_train_augmentation_model():
+    model = keras.Sequential(
+        [
+            layers.Rescaling(1 / 255.0),
+            layers.Resizing(INPUT_SHAPE[0] + 20, INPUT_SHAPE[0] + 20),
+            layers.RandomCrop(IMAGE_SIZE, IMAGE_SIZE),
+            layers.RandomFlip("horizontal"),
+        ],
+        name="train_data_augmentation",
+    )
+    return model
+
+
+def get_test_augmentation_model():
```

Comment:
This is just preprocessing, not augmentation. You should have 2 separate models, one for preprocessing, and one for augmentation (only used in training), since preprocessing is the same for both training and testing

Reply:
Thanks for the suggestion. Could you verify the changes?